### PR TITLE
feat(gherkin-onTrack): Implemented onTrack support for gherkin.

### PR DESCRIPTION
**Pull request by yasirfolio3**
# Summary
- Added support for on-track listener in gherkin.
- Updated clientwrapper struct properties to private.

# Test Result
https://travis-ci.com/optimizely/go-sdk/jobs/265499758

### CODE REVIEW

**Comment by yasirfolio3** [comment]
@mikeng13 Result:
https://travis-ci.com/optimizely/go-sdk/jobs/265499758

File: tests/integration/optlyplugins/test_composite_service.go [hunk 1]
```diff
@@ -35,15 +35,9 @@ func (c *TestCompositeService) AddListeners(listeners map[string]int) {
...
-	for listenerType, count := range listeners {
-		for i := 1; i <= count; i++ {
-			switch listenerType {
-			case "Decision":
-				c.OnDecision(c.decisionNotificationCallback)
-				break
-			default:
-				break
-			}
+	if count, ok := listeners[models.KeyDecision]; ok {
```
Comment (msohailhussain) [inline tests/integration/optlyplugins/test_composite_service.go]: is this change needed in `OnTrack` integration. if not please add proper description.

File: tests/integration/optlyplugins/track_listener_manager.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/optimizely/go-sdk/tests/integration/models"
+)
+
+// TrackListenerManager helper for track notification
+type TrackListenerManager struct {
+	listenersCalled []models.TrackListener
+}
+
+// TrackCallback represents callback function for track notification
+type TrackCallback = func(eventKey string, userContext entities.UserContext, eventTags map[string]interface{}, conversionEvent event.ConversionEvent)
```
Comment (msohailhussain) [inline tests/integration/optlyplugins/track_listener_manager.go]: am not in favor of adding more and more new `types` for every listener, I think only one `interface` is fineIf there is any issue lets discuss offline.

File: tests/integration/support/client_wrapper.go [hunk 3]
```diff
@@ -43,11 +43,12 @@ var sdkKey int
...
-	DecisionService    decision.Service
-	EventDispatcher    event.Dispatcher
-	UserProfileService decision.UserProfileService
-	OverrideStore      decision.ExperimentOverrideStore
+	client               *client.OptimizelyClient
+	decisionService      decision.Service
+	eventDispatcher      event.Dispatcher
+	userProfileService   decision.UserProfileService
+	overrideStore        decision.ExperimentOverrideStore
+	trackListenerManager *optlyplugins.TrackListenerManager
```
Comment (msohailhussain) [inline tests/integration/support/client_wrapper.go]: not needed then.

File: tests/integration/support/client_wrapper.go [hunk 4]
```diff
@@ -110,15 +111,19 @@ func GetInstance(apiOptions models.APIOptions) *ClientWrapper {
...
+	for _, callback := range trackListenerManager.GetListenerCallbacks(apiOptions.Listeners) {
+		client.OnTrack(callback)
+	}
 	clientInstance = &ClientWrapper{
-		Client:             client,
-		DecisionService:    decisionService,
-		EventDispatcher:    eventProcessor.EventDispatcher,
-		UserProfileService: userProfileService,
-		OverrideStore:      overrideStore,
+		client:               client,
```
Comment (msohailhussain) [inline tests/integration/support/client_wrapper.go]: Why this change is required in this PR. Lets have all these things in separate PR. This should focus only onTrack.

File: tests/integration/support/steps.go [hunk 5]
```diff
@@ -194,12 +193,19 @@ func (c *ScenarioCtx) InTheResponseKeyShouldBeObject(argumentType, value string)
 func (c *ScenarioCtx) InTheResponseShouldMatch(argumentType string, value *gherkin.DocString) error {
 	switch argumentType {
 	case models.KeyListenerCalled:
-		var requestListenersCalled []models.DecisionListener
-
-		if err := yaml.Unmarshal([]byte(value.Content), &requestListenersCalled); err != nil {
-			break
+		decisionListeners, trackListeners := getDecisionAndTrackListeners(value.Content)
```
Comment (msohailhussain) [inline tests/integration/support/steps.go]: let's refactor this code. we don't really need all these conditions.

**Comment by msohailhussain** [comment]
We don't really need separate TrackListener Manager, we can use only one `struct` for all Decision/`Track` and `LogEvent`

**Comment by msohailhussain** [comment]
looks good to me now. will take a look again and then will approve.

File: tests/integration/optlyplugins/notification_manager.go [hunk 6]
```diff
@@ -1,64 +1,43 @@
-/****************************************************************************
```
Comment (msohailhussain) [inline tests/integration/optlyplugins/notification_manager.go]: why it's removed please readd.

File: tests/integration/support/steps.go [hunk 7]
```diff
@@ -194,12 +193,8 @@ func (c *ScenarioCtx) InTheResponseKeyShouldBeObject(argumentType, value string)
 func (c *ScenarioCtx) InTheResponseShouldMatch(argumentType string, value *gherkin.DocString) error {
 	switch argumentType {
 	case models.KeyListenerCalled:
-		var requestListenersCalled []models.DecisionListener
-
-		if err := yaml.Unmarshal([]byte(value.Content), &requestListenersCalled); err != nil {
-			break
-		}
-		if subset.Check(requestListenersCalled, c.apiResponse.ListenerCalled) {
+		expectedListenersCalled := parseListeners(value.Content)
```
Comment (msohailhussain) [inline tests/integration/support/steps.go]: please add comment whats the purpose of parseListener

File: tests/integration/support/utils.go [hunk 8]
```diff
@@ -23,12 +23,37 @@ import (
...
 
+	"github.com/optimizely/go-sdk/tests/integration/models"
+
 	"github.com/google/go-cmp/cmp"
 	"github.com/optimizely/go-sdk/pkg/config"
 
 	"gopkg.in/yaml.v3"
 )
 
+func parseListeners(value string) (listeners []interface{}) {
```
Comment (msohailhussain) [inline tests/integration/support/utils.go]: Need to check if we can avoid this logic.

Comment (msohailhussain) [reply tests/integration/support/utils.go]: Discussed offline, we can avoid this parsing if comparison of types doesn't cause problem. 
When comparing two objects, yaml returns `[]interface{}` type while actual object returns `[]TrackListener`. My suggestion is remove `Listener` models and compare objects using `[]interface`. will try it out and update the code.

**Comment by msohailhussain** [comment]
ticket is added OASIS-5755 for parsing logic. @mikeng13

File: tests/integration/optlyplugins/notification_manager.go [hunk 9]
```diff
@@ -17,48 +17,43 @@
...
 	"github.com/optimizely/go-sdk/tests/integration/models"
 )
 
-// TestCompositeService represents a CompositeService with custom implementations
-type TestCompositeService struct {
-	decision.CompositeService
-	listenersCalled []models.DecisionListener
+// NotificationManager manager class for notification listeners
+type NotificationManager struct {
+	DecisionService *decision.CompositeService
```
Comment (mikeng13) [inline tests/integration/optlyplugins/notification_manager.go]: You don't need to call out the `DecisionService` separately. THis can be accessed via `n.Client.DecisionService`

File: tests/integration/optlyplugins/notification_manager.go [hunk 10]
```diff
@@ -17,48 +17,43 @@
...
+	Client          *client.OptimizelyClient
```
Comment (mikeng13) [inline tests/integration/optlyplugins/notification_manager.go]: Instead of retaining an instance of the client, maybe we can just pass it into `SubscribeNotifications`. I just think it's weird that the client wrapper has a reference to the client and this also has a reference to the client.
